[REF] ticketbai: update and fix fundamental functionality

blaskurain opened the pull request:
This consolidates the following PRs:

  - OCA/l10n-spain#1761
  - OCA/l10n-spain#1790
  - OCA/l10n-spain#1794
  - OCA/l10n-spain#1796
  - OCA/l10n-spain#1811
  - OCA/l10n-spain#1823
  - OCA/l10n-spain#1826
  - OCA/l10n-spain#1828
  - OCA/l10n-spain#1850

[IMP] Allow credit notes for invoices created before deploying ticketbai

[IMP] Allow the user to specify that the line detail is confidential

Gipuzkoa requires that invoices include line details but at the same
time requires that this details should be obfuscated in case the are
to be considered confidential for any reason (medical data for
example).

[FIX] Update URLs for Gipuzkoa

[FIX] Require country to be set for partners, don't default on "Spanish"

[REF] Ensure that no messages are sent to IRS from cloned databases

[UPD] Update Gipuzkoa Tax Agency Rest Api response codes

Update to the latest official documentation:
https://www.gipuzkoa.eus/documents/2456431/13761128/Anexo+IV.pdf/79d4cafc-eeaf-0b73-d4a9-3eb70c68034a

[UPD] Save warning messages from IRS also when invoices get accepted

[FIX] Update signature policy so that also Bizkaia accepts it.

enriquemartin commented on l10n_es_ticketbai/models/ticketbai_invoice.py:
```diff
@@ -44,3 +46,77 @@ def cancel_and_recreate(self):
+        help='Number prefix of this invoice name e.g. if the invoice name is'
+        ' INV/2021/00001 then the prefix is INV/2021/, '
+        'ending back slash included')
+    expedition_date = fields.Char(
```
Por qué no crear el campo de tipo `Date()` y luego formatearlo?

enriquemartin reviewed (approved):
Salvo el comentario del campo fecha, todo funciona según lo esperado.

xAdrianC-Kernet reviewed (approved):
Realizado ciclo de pruebas funcionales completo:

Factura de venta / cancelación / Rectificativa

Tanto en guipuzkoa como en bizkaia.

Todo ok

xAdrianC-Kernet commented:
hola @pedrobaeza este PR está probado y aunque tiene algunos push posteriores a los approves son cambios consensuados y probados, se podría mergear para que podamos seguir? 
Gracias.

pedrobaeza commented:
Pero aquí mezcla cambios de los existentes con lo nuevo de Batuz. Es demasiado para ir todo en un solo commit y además con demasiados pocos revisores, no crees? Desde luego, como mínimo para que esto sea razonable, diría que debe estar cada set lógico de cambios en su commit con su explicación. Vale que sea un único PR, pero que sea un commit, no.

blaskurain commented:
¡Vaya po Dios! Y yo aquí fusionando commits porque tenía entendido que se preferían PRs con un único commit. Intentaré recuperar los estados intermedios, a ver. @pedrobaeza, ¿serías tan amable de comentarme qué hay de cierto (si es que hay algo... la verdad es que no encuentro la referencia) en la impresión que tenía de que se prefieren PRs con un único commit, a ver si no tropiezo con la misma piedra otra vez?

xAdrianC-Kernet commented:
Yo creo que en este caso, basado en el conocimiento que tenemos los que estamos revisando y la dependencia de unos módulos con otros, es mejor todo en un PR y en un commit (siguiendo lo que se sugiere desde OCA y en otros hilos). Estos módulos por separado o no funcionan o no tienen la funcionalidad suficiente para ser probados.

pedrobaeza commented:
Siento la confusión. Partimos de que no existe la verdad absoluta. La cuestión es dividir por unidades lógicas los cambios, y que al final eso sirva para que cuando se mire atrás, podamos deducir el por qué de las cosas (gracias a mensajes de commit currados :wink: ) y saber qué se ha cambiado.

Esa impresión tal vez la tienes porque otras veces sobre un mismo bugfix, se hacen muchos commits porque van saliendo revisiones o demás, y claro, al final es refinar ese mismo bugfix y no tiene sentido mantenerlo por separado. Yo mismo soy un fanático de tener una historia de commits lo menor posible, pero tampoco sin tener commits "cajón-desastre".

Poniendo una analogía con las fotos, imagina que tienes fotos de tu cumpleaños que te han proporcionado varios amigos. Ponerlos cada uno en un album separado haría que fueran más difíciles de buscar y demás, y por eso se juntan siempre en el album "Mi cumpleaños". Pero si ahora esos amigos te proporcionan fotos de todos los cumpleaños y eventos del año, tenerlos todos juntos en el mismo album te hará que sea mucho más difícil localizar cualquier cosa o recordar si esa foto era de tal o cual evento. ¿Qué tal con esta analogía?

@xAdrianC-Kernet el PR está para aunar varios commits de un golpe y que al fusionar se tenga todo a la vez, pero eso no significa que tenga que estar todo en un commit.

blaskurain committed:
[ADD] l10n_es_ticketbai_api_batuz, l10n_es_ticketbai_api Add Batuz support.

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Never send messages to IRS from cloned databases, avoid manual commits.

xAdrianC-Kernet commented:
Perfecto @pedrobaeza voy a comentar con @blaskurain para separar los cambios

blaskurain committed:
[IMP] l10n_es_ticketbai_api Use custom signature policy for each tax agency.

blaskurain committed:
[REF] l10n_es_ticketbai_api Move XML schema dir to tests and make usage more flexible.

Move the schema directory into the tests directory because after all
that is where it is used and also travis prefers it this way.

Make the usage of the catalog a little more flexible as stepping stone
for related addons.

blaskurain committed:
[IMP] l10n_es_ticketbai_api Update Gipuzkoa Tax Agency Rest Api response codes and URLs

Update according to the latest official documentation ->
https://www.gipuzkoa.eus/documents/2456431/13761128/Anexo+IV.pdf/79d4cafc-eeaf-0b73-d4a9-3eb70c68034a

blaskurain committed:
[IMP] l10n_es_ticketbai Add support for "IVA Exento Repercutido Sujeto"

blaskurain committed:
[FIX] l10n_es_ticketbai_api If an invoice substitutes simplified invoice, it is not a refund

blaskurain committed:
[FIX] l10n_es_ticketbai_api Remove access to tbai.certificate for group_user.

Allowing this access could be a security risk.

blaskurain committed:
[IMP] l10n_es_ticketbai_api Save warning messages from IRS also when invoices get accepted

blaskurain committed:
[IMP] l10n_es_ticketbai Allow the user to say that the line detail is confidential

Gipuzkoa requires that invoices include line details but at the same
time requires that this details should be obfuscated in case the are
to be considered confidential for any reason (medical data for
example).

blaskurain committed:
[IMP] l10n_es_ticketbai Credit notes for invoices created before deploying ticketbai

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Add missing tests.

blaskurain committed:
[IMP] l10n_es_ticketbai Provide the TBAI testing environment only if available.

The testing environment is not available for all the tax agency. Try
not to lie to the user letting him activate the testing environment
when we know it is not available.

blaskurain committed:
[IMP] l10n_es_ticketbai simplify document schema validation.

Instead of raise + except just ask the schema whether a document is
valid: much simpler and cleaner. Also, keeps travis happy by not
logging assertion errors.

blaskurain committed:
[IMP] l10n_es_ticketbai_api: report error if partner has not country code

blaskurain commented:
Gracias @pedrobaeza por la aclaración. Acabo de subir el mismo código separado como dices en unidades lógicas como mejor he podido. La más grande es con diferencia todo el bloque de Batuz en los addons l10n_es_ticketbai_api_batuz, l10n_es_ticketbai_api que son nuevos completamente.

xAdrianC-Kernet reviewed (approved):
Realizadas todas las pruebas de nuevo, tanto para Guipuzkoa como Bizkaia, todo correcto.

pedrobaeza commented:
OK, fusionando:

/ocabot merge nobump
